torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -1096,13 +1098,18 @@ class TORCH_CUDA_API GridReduction final : public Expr {
     return sync_buffer_;
   }
 
+  const ParallelTypeBitmap& thread_pred() const {
+    return thread_pred_;
+  }
+
   static std::string getPredicateFlagName(const TensorView* val);
   static std::string getPredicateFlagName(const fuser::cuda::TensorView* val);
 
  private:
   ReductionOp* reduction_op_ = nullptr;
   Allocate* reduction_buffer_ = nullptr;
   Allocate* sync_buffer_ = nullptr;
+  ParallelTypeBitmap thread_pred_;
```

Comment:
Why can't we use the "normal" predicate (Expr::predicate_) instead of this?

Reply:
We could. We would need to make some changes to `gridReduce`. In particular, indexing the work buffer written by each thread block needs some non-significant change. 

However, `gridReduce` does have template bool parameters exactly for predicating based on block and thread indices being zero, so using those template flags should make more sense. The normal predicate is not for thread/block indices, so it can't be mapped to the template flags, and that's why we need to separate them for `gridReduce`. Note that for other expressions, we just combine them by `&&`.

Reply:
Added a comment to the code itself too.